redbot/core/core_commands.py

```diff
@@ -1940,6 +1941,357 @@ async def reload(self, ctx: commands.Context, *cogs: str):
+            cog_commands[module].append((command.name, command.type, True))
+            if command.type is discord.AppCommandType.message:
+                message_command_names.add(command.name)
+            elif command.type is discord.AppCommandType.user:
+                user_command_names.add(command.name)
+        for command in self.bot.tree._disabled_context_menus.values():
+            module = command.module
+            if "." in module:
+                module = module[: module.find(".")]
+            cog_commands[module].append((command.name, command.type, False))
+
+        # Commands added with evals will come from __main__, make them unknown instead
+        if "__main__" in cog_commands:
+            main_data = cog_commands["__main__"]
+            del cog_commands["__main__"]
+            cog_commands["(unknown)"] = main_data
+
+        # Commands enabled but unloaded won't appear unless accounted for
+        enabled_commands = await self.bot.list_enabled_app_commands()
+        unknown_slash = set(enabled_commands["slash"]) - slash_command_names
+        unknown_message = set(enabled_commands["message"]) - message_command_names
+        unknown_user = set(enabled_commands["user"]) - user_command_names
+
+        unknown_slash = [(n, discord.AppCommandType.chat_input, True) for n in unknown_slash]
+        unknown_message = [(n, discord.AppCommandType.message, True) for n in unknown_message]
+        unknown_user = [(n, discord.AppCommandType.user, True) for n in unknown_user]
+
+        cog_commands["(unknown)"].extend(unknown_slash)
+        cog_commands["(unknown)"].extend(unknown_message)
+        cog_commands["(unknown)"].extend(unknown_user)
+        # Hide it when empty
+        if not cog_commands["(unknown)"]:
+            del cog_commands["(unknown)"]
+
+        if not cog_commands:
+            await ctx.send(_("There are no application commands to list."))
+            return
+
+        msg = ""
+        for cog in sorted(cog_commands.keys()):
+            msg += cog + "\n"
+            for name, raw_command_type, enabled in sorted(cog_commands[cog], key=lambda v: v[0]):
+                diff = "+ " if enabled else "- "
+                command_type = "unknown"
+                if raw_command_type is discord.AppCommandType.chat_input:
+                    command_type = "slash"
+                elif raw_command_type is discord.AppCommandType.message:
+                    command_type = "message"
+                elif raw_command_type is discord.AppCommandType.user:
+                    command_type = "user"
+                msg += diff + command_type.ljust(7) + " | " + name + "\n"
+            msg += "\n"
+
+        pages = pagify(msg, delims=["\n\n", "\n"])
+        pages = [box(page, lang="diff") for page in pages]
+        await menu(ctx, pages)
+
+    @slash.command(name="sync")
+    @commands.cooldown(1, 60)
+    async def slash_sync(self, ctx: commands.Context, guild: discord.Guild = None):
```

Comment:
It might be worth wrapping the content of this command in `ctx.typing()` so the user knows something is happening immediately and then return a message instead of just ticking the original message with a reaction. We could go so far as to provide the number of commands synced so the user has a direct reference to where they're at on the command limit instead of guessing.